### PR TITLE
test(rln-relay): setup heartbeat ws subscription and log when connection drops

### PR DESCRIPTION
**Pull request by rymnc**
Does 3 things:
- Logs out when eth node connections are dropped
- Keeps track of last seen block so that when the connection is reinitiated the tree is not repopulated
- Heartbeat newHead subscription

In a follow up PR:
- Reconnection logic

### CODE REVIEW

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 1]
```diff
@@ -906,24 +906,9 @@ proc addAll*(wakuRlnRelay: WakuRLNRelay, list: seq[IDCommitment]): RlnRelayResul
...
-  web3.defaultAccount = ethAccountAddress 
-  #  set the gas price twice the suggested price in order for the fast mining
-  # let gasPrice = int(await web3.provider.eth_gasPrice()) * 2
-
-  # subscribe to the MemberRegistered events
-  # TODO can do similarly for deletion events, though it is not yet supported
-  discard await contractObj.subscribe(MemberRegistered, %*{"fromBlock": blockNumber, "address": contractAddress}) do(pubkey: Uint256, index: Uint256){.raises: [Defect], gcsafe.}:
+proc subscribeToMemberRegistrations(web3: Web3, contractAddress: Address, handler: RegistrationEventHandler): Future[Subscription] {.async, gcsafe} =
+  var contractObj = web3.contractSender(MembershipContract, contractAddress)
+  return await contractObj.subscribe(MemberRegistered, %*{"fromBlock": "0x0", "address": contractAddress}) do(pubkey: Uint256, index: Uint256){.raises: [Defect], gcsafe.}:
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim:911]: the try-except is part of the callback and errorHandler procs passed into the subscription proc. They will get executed regardless :)

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 2]
```diff
@@ -906,24 +906,9 @@ proc addAll*(wakuRlnRelay: WakuRLNRelay, list: seq[IDCommitment]): RlnRelayResul
...
-  # prepare a contract sender to interact with it
-  var contractObj = web3.contractSender(MembershipContract, contractAddress) 
-  web3.defaultAccount = ethAccountAddress 
-  #  set the gas price twice the suggested price in order for the fast mining
-  # let gasPrice = int(await web3.provider.eth_gasPrice()) * 2
-
-  # subscribe to the MemberRegistered events
-  # TODO can do similarly for deletion events, though it is not yet supported
-  discard await contractObj.subscribe(MemberRegistered, %*{"fromBlock": blockNumber, "address": contractAddress}) do(pubkey: Uint256, index: Uint256){.raises: [Defect], gcsafe.}:
+proc subscribeToMemberRegistrations(web3: Web3, contractAddress: Address, handler: RegistrationEventHandler): Future[Subscription] {.async, gcsafe} =
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim]: Addressed in [fc676c1](https://github.com/status-im/nwaku/pull/1253/commits/fc676c10763cf3b7efe24a9e63bcaa42e70b2b79)

**Comment by rymnc** [comment]
> LGTM! Just to make valid roots slightly more human-readable, can you change
> 
> * Line 827:
> 
> ```
> debug "invalid message: provided root does not belong to acceptable window of roots", provided=msg.proof.merkleRoot, validRoots=rlnPeer.validMerkleRoots.mapIt("0x" & (it.toHex))
> ```
> 
> * Line 1124:
> 
> ```
> debug "acceptable window", validRoots=rlnPeer.validMerkleRoots.mapIt("0x" & (it.toHex))
> ```

Addressed in [db679a5](https://github.com/status-im/nwaku/pull/1253/commits/db679a5ff6abdd99842391a1e2010b662e2aecda)!

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 3]
```diff
@@ -824,8 +824,7 @@ proc validateMessage*(rlnPeer: WakuRLNRelay, msg: WakuMessage,
 
   ## TODO: FIXME after resolving this issue https://github.com/status-im/nwaku/issues/1247
   if not rlnPeer.validateRoot(msg.proof.merkleRoot):
-    debug "invalid message: provided root does not belong to acceptable window of roots", provided=msg.proof.merkleRoot, validRoots=rlnPeer.validMerkleRoots
-    waku_rln_invalid_messages_total.inc(labelValues=["invalid_root"])
+    debug "invalid message: provided root does not belong to acceptable window of roots", provided=msg.proof.merkleRoot, validRoots=rlnPeer.validMerkleRoots.mapIt("0x" & (it.toHex))    waku_rln_invalid_messages_total.inc(labelValues=["invalid_root"])
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim]: Addressed!

**Review by jm-clius — APPROVED** [review]
LGTM!

**Comment by jm-clius** [comment]
A general comment for the RLN module: although we certainly don't follow this consistently everywhere, [NEP-1](https://nim-lang.org/docs/nep1.html) suggests a max line length of 80 characters. Perhaps it's possible to use multiline formatting with indenting more regularly to improve readability. It also makes sense to keep log messages _somewhat_ succinct, as most log formatters would be much more readable with shorter lines.

**Comment by rymnc** [comment]
> A general comment for the RLN module: although we certainly don't follow this consistently everywhere, [NEP-1](https://nim-lang.org/docs/nep1.html) suggests a max line length of 80 characters. Perhaps it's possible to use multiline formatting with indenting more regularly to improve readability. It also makes sense to keep log messages _somewhat_ succinct, as most log formatters would be much more readable with shorter lines.

Agreed, will add to the refactoring list!

**Comment by rymnc** [comment]
Merging since the `wakuv2.test` fleet is using this branch, and no functionality has been changed.

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 4]
```diff
@@ -934,6 +919,32 @@ proc subscribeToGroupEvents(ethClientUri: string, ethAccountAddress: Address, co
...
+  ## it collects all the events starting from the given `blockNumber`
+  ## for every received event, it calls the `handler`
+  let web3 = await newWeb3(ethClientUri)
+  var latestBlock: Quantity
+  let newHeadCallback = proc (blockheader: BlockHeader) {.gcsafe.} =
+    latestBlock = blockheader.number
+    debug "block received", blockNumber = latestBlock
+  let newHeadErrorHandler = proc (err: CatchableError) {.gcsafe.} =
+    error "Error from subscription: ", err=err.msg
+  discard await web3.subscribeForBlockHeaders(newHeadCallback, newHeadErrorHandler)
```
Comment (staheri14) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim:934]: This subscription is handled separately from the subscription to the `MemberRegistered` events on the contract. The `latestBlock` perceived by `subscribeForBlockHeaders` is not necessarily the same as the same block processed in subscription to the `MemberRegistered` events (they are two separate queries and streams of data with no correlation). Have you checked that?

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 5]
```diff
@@ -934,6 +919,32 @@ proc subscribeToGroupEvents(ethClientUri: string, ethAccountAddress: Address, co
   do (err: CatchableError):
     error "Error from subscription: ", err=err.msg
 
+proc subscribeToGroupEvents(ethClientUri: string, ethAccountAddress: Address, contractAddress: Address, blockNumber: string = "0x0", handler: RegistrationEventHandler) {.async, gcsafe.} = 
+  ## connects to the eth client whose URI is supplied as `ethClientUri`
+  ## subscribes to the `MemberRegistered` event emitted from the `MembershipContract` which is available on the supplied `contractAddress`
+  ## it collects all the events starting from the given `blockNumber`
+  ## for every received event, it calls the `handler`
```
Comment (staheri14) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim:926]: When the internals of a proc changes, please also change the proc description to reflect the most recent state.

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim (hunk 4 above)
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim:934]: Yes, I mentioned this in our call. When we actually process blocks atomically, I will add another field to the WakuRlnRelay object called `lastProcessedBlock`. `lastSeenBlock` will always be an indicator of the health of the websocket connection.

**Pull request merged**